### PR TITLE
Replace built-in metrics by custom instrumentation middleware

### LINKED ISSUE #5943: Replace built-in metrics by custom instrumentation middleware
Recently, we discovered that vendoring a customized version of our instrumentation middleware offers significant advantages, primarily due to the added flexibility it provides for managing metric attributes. Upstream instrumentation libraries lack options for hooking into metrics to modify their behavior. These metrics often include attributes that do not apply to standard use cases. Additionally, contributing upstream seems to be a bit challenging, as spans and metrics are tightly coupled in their implementation.

The goal of this task is to:
1. remove the opentelemetry-instrumentation-* dependencies and switch to the instrumentation-sdk only,
2. introduce a custom middleware for api-app and content-app, optionally, for worker-app,
3. remove the current testing machinery for traces and metrics and replace it with reasonable unit tests,
4. introduce a new setting that activates the opentelemetry instrumentation on the application's side.

We already have a working concept in the pulp-service repository:
content-app: https://github.com/pulp/pulp-service/blob/4a72206d961b0d3e751405850b27bdc366414682/pulp_service/pulp_service/app/content.py#L69
api-app: https://github.com/pulp/pulp-service/blob/4a72206d961b0d3e751405850b27bdc366414682/pulp_service/pulp_service/app/middleware.py#L74

### PR DESCRIPTION
**Pull request by lubosmj**
closes #5943

### CODE REVIEW

File: pulpcore/tests/unit/metrics/test_aiohttp_metrics.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    joined_output = "".join(output.readlines())
+    result_0 = json.loads(joined_output.strip(os.linesep))
+
+    assert len(result_0) > 0
+
+    metrics = result_0["resource_metrics"][0]["scope_metrics"][0]
+    point_attrs = metrics["metrics"][0]["data"]["data_points"][0]["attributes"]
+    assert point_attrs["http.status_code"] == "2xx"
+    assert point_attrs["http.method"] == "GET"
+    assert point_attrs["http.route"] == "noop_response"
```
Comment (bmbouter) [inline pulpcore/tests/unit/metrics/test_aiohttp_metrics.py:44]: :smile:

**Review by bmbouter — DISMISSED** [review]
This looks good to me and also like what we talked about at the last pulpcore meeting. If anyone wants this done differently or has thoughts on this code please do let us know. Thanks!

File: pulpcore/tests/unit/metrics/test_aiohttp_metrics.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import json
+import os
+import pytest
+
+from aiohttp import web
+from io import StringIO
```
Comment (bmbouter) [inline pulpcore/tests/unit/metrics/test_aiohttp_metrics.py:6]: I would have used `mock` here, but this is fine.

File: pulpcore/tests/unit/metrics/test_django_metrics.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import json
+import os
+
+from io import StringIO
```
Comment (bmbouter) [inline pulpcore/tests/unit/metrics/test_django_metrics.py:4]: I would have used mock here, but this is fine.

File: pulpcore/app/util.py [hunk 4]
```diff
@@ -29,6 +36,30 @@
 from pulpcore.exceptions.validation import InvalidSignatureError
 
 
+@lru_cache(maxsize=1)
+def init_otel_meter(service_name, exporter=None, reader=None, provider=None):
```
Comment (decko) [inline pulpcore/app/util.py:40]: What if we have a telemetry module/package? Like, a `telemetry.py` file where we concentrate all those things?

File: pulpcore/tests/unit/metrics/test_aiohttp_metrics.py (hunk 2 above)
Comment (decko) [reply pulpcore/tests/unit/metrics/test_aiohttp_metrics.py:6]: Maybe the https://opentelemetry-python.readthedocs.io/en/latest/sdk/metrics.export.html#opentelemetry.sdk.metrics.export.InMemoryMetricReader

Comment (lubosmj) [reply pulpcore/tests/unit/metrics/test_aiohttp_metrics.py]: Hmm, I was mainly inspired by https://github.com/open-telemetry/opentelemetry-python/blob/897e8c93d11e7b6f26560d6f5329e506bb04b04a/opentelemetry-sdk/tests/metrics/integration_test/test_console_exporter.py#L99. I think we can leave it as it is for now.

**Review by decko — DISMISSED** [review]
LTTM 🚀

**Comment by lubosmj** [comment]
I am going to update the tests as suggested. In addition to that, I going to figure out how to stop emitting metrics in case the collector is improperly configured. Interestingly, this is visible only locally.
```
pulp_1   | During handling of the above exception, another exception occurred:
pulp_1   | 
pulp_1   | Traceback (most recent call last):
pulp_1   |   File "/usr/local/lib/python3.9/site-packages/opentelemetry/sdk/metrics/_internal/export/__init__.py", line 541, in _receive_metrics
pulp_1   |     self._exporter.export(
pulp_1   |   File "/usr/local/lib/python3.9/site-packages/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py", line 203, in export
pulp_1   |     resp = self._export(serialized_data.SerializeToString())
pulp_1   |   File "/usr/local/lib/python3.9/site-packages/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py", line 173, in _export
pulp_1   |     return self._session.post(
pulp_1   |   File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 637, in post
pulp_1   |     return self.request("POST", url, data=data, json=json, **kwargs)
pulp_1   |   File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 589, in request
pulp_1   |     resp = self.send(prep, **send_kwargs)
pulp_1   |   File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 703, in send
pulp_1   |     r = adapter.send(request, **kwargs)
pulp_1   |   File "/usr/local/lib/python3.9/site-packages/requests/adapters.py", line 700, in send
pulp_1   |     raise ConnectionError(e, request=request)
pulp_1   | requests.exceptions.ConnectionError: HTTPConnectionPool(host='localhost', port=4318): Max retries exceeded with url: /v1/metrics (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f8ecb13a8b0>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

**Comment by lubosmj** [comment]
There is no real way to tell the SDK to stop exporting data in case the collector does not exist. Logs yelling at users should be visible enough to suggest consulting the documentation.

File: pulpcore/app/util.py (hunk 4 above)
Comment (decko) [reply pulpcore/app/util.py]: Also, why you use `maxsize=1` here?

**Review by decko — APPROVED** [review]
LTTM 🚀

**Pull request merged**